src/main/scala/reporting/Converter.scala

```diff
@@ -0,0 +1,531 @@
+          case e: ExtractedModelEntry => ConstantEntry(e.toString)
+        }
+        getFunctionValue(model, fname, argsFinal, toSort)
+
+      case Combine(p0, p1) =>
+        //assuming Combine can only contain other snap.combine and snap.unit
+        val p0Eval = evaluateTerm(p0, model).asValueEntry
+        val p1Eval = evaluateTerm(p1, model).asValueEntry
+        val entry = ApplicationEntry("$Snap.combine", Seq(p0Eval, p1Eval))
+        UnprocessedModelEntry(entry)
+
+      case First(p) =>
+        val sub = evaluateTerm(p, model)
+        sub match {
+          case UnprocessedModelEntry(ApplicationEntry(name, args)) =>
+            if (name == "$Snap.combine") {
+              UnprocessedModelEntry(args.head)
+            } else {
+              OtherEntry(s"First($p)", "unapplicable")
+            }
+          case OtherEntry(t, _) => OtherEntry(s"First($t)", "unapplicable")
+          case _                => OtherEntry(s"First($sub)", "unapplicable")
+        }
+
+      case Second(p) =>
+        val sub = evaluateTerm(p, model)
+        sub match {
+          case UnprocessedModelEntry(ApplicationEntry(name, args)) =>
+            if (name == "$Snap.combine") {
+              UnprocessedModelEntry(args(1))
+            } else {
+              OtherEntry(s"Second($p})", "unapplicable")
+            }
+          case OtherEntry(t, _) => OtherEntry(s"Second($t)", "unapplicable")
+          case _                => OtherEntry(s"Second($sub)", "unapplicable")
+        }
+
+      case SortWrapper(t, to) =>
+        val sub = evaluateTerm(t, model)
+        val fromSortName: String = translateSort(t.sort)
+        val toSortName: String = translateSort(to)
+        val fname = s"$$SortWrappers.${fromSortName}To$toSortName"
+
+        sub match {
+          case UnprocessedModelEntry(entry) =>
+            getFunctionValue(model, fname, Seq(entry), to)
+          case OtherEntry(t, _) =>
+            OtherEntry(s"SortWrapper($t)", "unapplicable")
+          case _ => OtherEntry(s"SortWrapper($t)", "unapplicable")
+        }
+
+      case PredicateLookup(predname, psf, args) =>
+        val lookupFuncName: String = s"$$PSF.lookup_$predname"
+        val snap = toSnapTree.apply(args)
+        val snapVal = evaluateTerm(snap, model).asValueEntry
+        val psfVal = evaluateTerm(psf, model).asValueEntry
+        val arg = Seq(psfVal, snapVal)
+        val result = getFunctionValue(model, lookupFuncName, arg, sorts.Snap)
+        result
+
+      case _ =>
+        OtherEntry(s"$term", "unhandled")
+    }
+  }
+
+  def extractHeap(h: Iterable[Chunk], model: Model): ExtractedHeap = {
+    var entries: List[HeapEntry] = List()
+    h foreach {
+      case c @ BasicChunk(FieldID, _, _, _, _) =>
+        val entry = extractField(c, model)
+        entries = entries :+ entry
+      case c @ BasicChunk(PredicateID, _, _, _, _) =>
+        val entry = extractPredicate(c, model)
+        entries = entries :+ entry
```

Comment:
Probably doesn't matter much in practice in this case, but constructing a linked-list by repeatedly appending to its end is the worst-possible way w.r.t performance. A vector might be more suited here (see e.g. https://docs.scala-lang.org/overviews/collections-2.13/introduction.html).